subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java

```diff
@@ -79,11 +76,6 @@ public class PublicApiCorrectnessTest {
                 .or(type(QName.class))
                 .or(type(BiFunction.class))
                 .as("built-in JDK classes"))
-            .or(type(Node.class)
-                .or(type(MarkupBuilder.class))
-                .or(type(Closure.class))
-                .as("Groovy classes")
-            )
             .or(type(Function1.class)
```

Comment:
❓ Should this ADR also comment on these Kotlin types?  

The Kotlin `Pair` class being here seems especially suspicious.  Why would we need this instead of a) Using a different class already in the JDK, b) using Pair from Apache Commons or some other popular library, c) promoting `org.gradle.internal.Pair` to public API?

Reply:
+1 I think the ADR should be about discouraging any language-specific types in the public API, not just Groovy.

Reply:
I mostly agree, but I think this needs to be handled separately for a few reasons:
- The "public API" here isn't purely the public API. It also contains the Kotlin DSL extensions.
- The Kotlin DSL extensions require the Kotlin types and we don't want to just accept them as exceptions to the test.
- People aren't introducing Kotlin types in the classes written in Java, but they _are_ adding Groovy types because they're copying what's already there. 

I think this means we need...
- A set of public API methods that include everything
- A set of Kotlin DSL public methods
- A set of Groovy DSL public methods (if these exist?)
- A set of public API methods that exclude the language specific packages/types

And then we can enforce rules that apply to all methods (most rules), Kotlin DSL specific methods (allow Kotlin specific types), Groovy DSL specific methods (allow Groovy specific types) and non-DSL specific methods (disallow Groovy and Kotlin types with exceptions).

It looks like `org.gradle.kotlin.dsl` might have all the public types that need Kotlin types.